File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Nodejs/core/interpreter/Interpreter.js
const { Handler } = require('../handler/Handler')
const CommandSerializer = require('../protocol/CommandSerializer')
const CommandDeserializer = require('../protocol/CommandDeserializer')
const Runtime = require("../../utils/RuntimeName");
const ConnectionType = require('../../utils/ConnectionType');

let Transmitter
let Receiver

class Interpreter {
    handler = new Handler()

    async executeAsync(command, connectionData) {
        try {
            const messageByteArray = new CommandSerializer().serialize(command, connectionData)

            if (connectionData.connectionType === ConnectionType.WEB_SOCKET) {
                const { WebSocketClient } = require("../webSocketClient/WebSocketClient");
                const wsClient = new WebSocketClient(connectionData.hostname, null)

                const responseByteArray = await wsClient.send(messageByteArray)
                return new CommandDeserializer(responseByteArray).deserialize()
            }
            else {
                return this.execute(command, connectionData)
            }
        } catch (error) {
            // TODO: handle error exception log?
        }
    }
    
    execute(command, connectionData) {
        try {
            let messageByteArray = new CommandSerializer().serialize(command, connectionData)
            let responseByteArray
            
            if (command.runtimeName === Runtime.Nodejs && connectionData.connectionType === ConnectionType.IN_MEMORY)
            {
                // lazy receiver loading
                if (!Receiver) {
                    Receiver = require('../receiver/Receiver')
                }
                responseByteArray = Receiver.sendCommand(messageByteArray)

            }
            else {
                // lazy transmitter loading
                if (!Transmitter) {
                    Transmitter = require('../transmitter/Transmitter')
                }
                responseByteArray = Transmitter.sendCommand(messageByteArray)
            }
            return new CommandDeserializer(responseByteArray).deserialize()
        } catch (error) {
            // TODO: handle error exception log?
        }
    }

    process(messageByteArray) {
        let receivedCommand = new CommandDeserializer(messageByteArray).deserialize()
        return this.handler.handleCommand(receivedCommand)
    }
}

module.exports = Interpreter

File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Python/javonet/core/interpreter/Interpreter.py
from javonet.core.handler.Handler import Handler
from javonet.core.protocol.CommandDeserializer import CommandDeserializer
from javonet.core.protocol.CommandSerializer import CommandSerializer
from javonet.utils.ConnectionType import ConnectionType
from javonet.utils.RuntimeName import RuntimeName
from javonet.utils.connectionData.IConnectionData import IConnectionData

handler = Handler()

class Interpreter:

    def execute(self, command, connection_data: IConnectionData):
        message_byte_array = CommandSerializer().serialize(command, connection_data)
        if connection_data.connection_type == ConnectionType.WebSocket:
            from javonet.core.webSocketClient.WebSocketClient import WebSocketClient
            response_byte_array = WebSocketClient().send_message(connection_data, message_byte_array)
        elif (command.runtime_name == RuntimeName.python) & (
                connection_data.connection_type == ConnectionType.InMemory):
            from javonet.core.receiver.Receiver import Receiver
            response_byte_array = Receiver().SendCommand(message_byte_array, len(message_byte_array))
        else:
            from javonet.core.transmitter.Transmitter import Transmitter
            response_byte_array = Transmitter.send_command(message_byte_array)

        return CommandDeserializer(response_byte_array).deserialize()

    def process(self, message_byte_array):
        received_command = CommandDeserializer(message_byte_array).deserialize()
        return handler.handle_command(received_command)
